software/contrib/quantizer.py

```diff
@@ -0,0 +1,495 @@
+            if self.scale[note]:
+                delta = abs(nearest_semitone - note)
+                if delta < best_delta:
+                    nearest_on_scale = note
+                    best_delta = delta
+            
+        self.current_note = nearest_on_scale
+        self.output_voltage = base_volts + (self.root + nearest_on_scale) * VOLTS_PER_SEMITONE + self.octave * VOLTS_PER_OCTAVE
+        
+    ## Read the input signal, quantize it, set outputs 1-5 accordingly
+    #
+    #  Called by the main loop in continuous mode or the rising clock handler
+    #  in triggered mode
+    def read_quantize_output(self):
+        self.input_voltage = ain.read_voltage(128)
+        self.quantize(self.input_voltage)
+        
+        cv1.voltage(self.output_voltage)
+        
+        for i in range(len(self.aux_outs)):
+            self.aux_outs[i].voltage(self.output_voltage + self.intervals[i] * VOLTS_PER_SEMITONE)
+    
+    ## Handler for the rising edge of the input clock
+    def on_trigger(self):
+        if self.mode == MODE_TRIGGERED:
+            self.read_quantize_output()
+            cv6.on()
+            
+    ## Handler for the falling edge of the input clock
+    def after_trigger(self):
+        if self.mode == MODE_TRIGGERED:
+            cv6.off()
+            
+    ## Handler for pressing button 1
+    #
+    #  Button 1 is used for the main interaction and is passed to
+    #  the current display for user interaction
+    def on_button1(self):
+        self.last_interaction_time = time.time()
+        self.active_screen.on_button1()
+        
+    ## Handler for pressing button 2
+    #
+    #  Button 2 is used to cycle between screens
+    def on_button2(self):
+        self.last_interaction_time = time.time()
+        
+        if self.active_screen == self.kb:
+            self.active_screen = self.menu
+        else:
+            self.active_screen = self.kb
+    
+    ## The main loop
+    #
+    #  Connects event handlers for clock-in and button presses
+    #  and runs the main loop
+    def main(self):
+        # connect the trigger handler here instead of the constructor
+        # otherwise it will start quantizing as soon as we instantiate the class
+        @din.handler
+        def on_rising_clock():
+            self.on_trigger()
+            
+        @din.handler_falling
+        def on_falling_clock():
+            self.after_trigger()
+            
+        @b1.handler
+        def on_b1_press():
+            self.on_button1()
+            
+        @b2.handler
+        def on_b2_press():
+            self.on_button2()
+        
+        while True:
+            # Update at 100Hz
+            CYCLE_RATE = 0.01
+            time.sleep(CYCLE_RATE)
+            
+            # Check if we've been idle for too long; if so, blank the screen
+            # to prevent burn-in
+            now = time.time()
+            if now - self.last_interaction_time > SCREENSAVER_TIMEOUT_S:
```

Comment:
[required] Sorry I missed this the first time, MicroPython is a little funny with time math. The recommended usage is [ticks_diff](https://docs.micropython.org/en/latest/library/time.html#time.ticks_diff) because the pico internal clock could wrap around which would cause this math to fail unexpectedly.

Reply:
Fair enough. This should be fixed in 631928a

Reply:
...er, 80262fc.  It's important to make sure the file is saved before you commit...

Reply:
Thanks for the quick fix! Let's also use the `ticks_diff` and use `time_ms` to make sure we're using correct values in the logic as well as when we request the current time. I also suggest updating the const name to use the `MS` suffix to match the unit of time. (Double check my work, I'm _pretty_ sure  this is the correct order of values in ticks_diff)

```suggestion
            if time.ticks_diff(self.last_interaction_time, time.now_ms()) > SCREENSAVER_TIMEOUT_MS:
```
Where `SCREENSAVER_TIMEOUT_MS ` is:
```python
SCREENSAVER_TIMEOUT_MS = 1000 * 60 * 20
```

Reply:
Ha, async overlapping comments :)

Double check the order of time values in your call to `ticks_diff`. I believe you might need to reverse them according to the docs examples for "Scheduling Events" - https://docs.micropython.org/en/latest/library/time.html#time.ticks_diff

Reply:
Just double-checked with a 10s timeout. The order in the latest commit does work as intended; after 10s of being idle the screen shuts off, and wakes back up on a button press.  I'm testing again with the full 20-minute timeout, but I don't expect it to behave differently.

I based it on the timeout example:
```
# Wait for GPIO pin to be asserted, but at most 500us
start = time.ticks_us()
while pin.value() == 0:
    if time.ticks_diff(time.ticks_us(), start) > 500:
        raise TimeoutError
```
which is equivalent to shutting the screen off after the timeout elapses:
```
now = time.ticks_us()
if time.ticks_diff(now, self.last_interaction_time) > SCREENSAVER_TIMEOUT_US:
    self.active_screen = self.screensaver
```
where the current time is the same and `start` is equivalent to `last_interaction_time`

Reply:
There was a bug with the 20 minute timeout.  Looks like there may have been an integer overflow when dealing with intervals that big.  Switching to ms seems to have fixed it though.

Reply:
Thanks for confirming and testing. LGTM.